Add deprecation notice for BS2 theme (#1642)

anvit opened the pull request:
Add deprecation notice that informs users that Bootstrap 2 themes are deprecated and will not be supported in future releases.

anvit committed:
Add deprecation notice for BS2 theme (#1642)

Add a notice informing users about BS2 themes being deprecated and
pointing them to a docs page that tells them how to switch themes.

anvit commented:
@melaniekung, @jraddaoui Still need to link this to an actual docs page, but I wanted your feedback on this, especially about if [apps/qubit/templates/_header.php](https://github.com/artefactual/atom/blob/c088704e63506d6bca5d9ea526e04fe59f12c5f9/apps/qubit/templates/_header.php#L1) is a good place to add this or if there's a better place to put it that might make client upgrades easier.

More details about what this looks like, and who will see this on the [issue page](https://github.com/artefactual/atom/issues/1642).

jraddaoui commented on apps/qubit/modules/default/actions/bs2DeprecationMessageComponent.class.php:
```diff
@@ -0,0 +1,40 @@
+{
+    public function execute($request)
+    {
+        if (!$this->context->user->isAuthenticated() || sfConfig::get('app_b5_theme', false)) {
```
I'd only show this to admins, like the update check component: https://github.com/artefactual/atom/blob/qa/2.x/apps/qubit/modules/default/actions/updateCheckComponent.class.php#L29.

anvit replied:
We'd talked about who should see it on the issue [here](https://github.com/artefactual/atom/issues/1642) and the thought was that regular authenticated users, especially editors or translators, would be likely to come across this before an admin does and could bring it to their attention.

jraddaoui replied:
Added a comment there ;)

jraddaoui commented on apps/qubit/modules/default/actions/bs2DeprecationMessageComponent.class.php:
```diff
@@ -0,0 +1,40 @@
+            null !== $this->context->user->getAttribute('bs2_deprecation_message_dismissed')
+        ) {
+            return sfView::NONE;
+        }
```
Nitpick: all conditions could go in the same if.

jraddaoui commented on apps/qubit/modules/default/templates/_bs2DeprecationMessage.php:
```diff
@@ -0,0 +1,12 @@
+<div class="animateNicely" id="bs2-deprecation-message">
+    <div class="alert alert-danger alert-banner">
+        <div id="bs2-deprecation-message-content">
+            <?php echo __($notificationMessage); ?>
```
Is this adding double translation? Is that an issue? I'd avoid the variable and use the string directly here.

jraddaoui commented on js/bs2DeprecationMessage.js:
```diff
@@ -0,0 +1,32 @@
+  class Bs2DeprecationMessage {
+    constructor(element) {
+      this.$element = $(element);
+      this.$Bs2DeprecationMessageBlock = this.$element.find('div[id="bs2-deprecation-message"]');
```
Nitpick: `this.$element.find('#bs2-deprecation-message');`

jraddaoui commented on js/bs2DeprecationMessage.js:
```diff
@@ -0,0 +1,32 @@
+    if (0 < $node.length)
+    {
+      new Bs2DeprecationMessage($node.get(0));
+    }
```
I never liked this class/prototype initialization from the body, you could fin the element here directly, `var $node = $('#bs2-deprecation-message');`. This will avoid the initialization when it's not there and the extra find in the constructor.

In general, I think a class/prototype is not necessary in these cases but I know that's what we use to do. I'm not sure if we could use Bootstrap events in this version, but check the privacy message JS in B5: https://github.com/artefactual/atom/blob/qa/2.x/plugins/arDominionB5Plugin/js/privacyMessage.js.

jraddaoui commented on js/bs2DeprecationMessage.js:
```diff
@@ -0,0 +1,32 @@
+    constructor(element) {
+      this.$element = $(element);
+      this.$Bs2DeprecationMessageBlock = this.$element.find('div[id="bs2-deprecation-message"]');
+      this.$Bs2DeprecationMessageButton = this.$element.find('button.bs2-deprecation-message-button');
```
Nitpick: `bs2-deprecation-message-button` looks like it should be an id instead of a class.

jraddaoui replied:
I think this was left behind ;)

anvit replied:
I've updated this to an ID, but I'm still using the class for styling. Let me know if our practice is to get rid of the class altogether in cases like this?

jraddaoui replied:
I think a class adds no value if it's only used by a single element and this one seems to specific to be reused. Thinking a bit more, I'd not even use an id in this case, being the only button inside `#bs2-deprecation-message`.

anvit replied:
Fixed!

jraddaoui replied:
Just nitpicking, I'd remove the class entirely and use the button in the LESS code too, nested under `#bs2-deprecation-message`.

jraddaoui commented on js/bs2DeprecationMessage.js:
```diff
@@ -0,0 +1,32 @@
+    constructor(element) {
+      this.$element = $(element);
+      this.$Bs2DeprecationMessageBlock = this.$element.find('div[id="bs2-deprecation-message"]');
+      this.$Bs2DeprecationMessageButton = this.$element.find('button.bs2-deprecation-message-button');
```
Nitpick: quite verbose variable names in a component already called the same, I'd name them `this.$block` and `this.$button`.

jraddaoui commented on plugins/arDominionPlugin/css/less/scaffolding.less:
```diff
@@ -1668,6 +1668,51 @@ body.login #content {
+  }
+}
+
+#bs2-deprecation-message {
+    text-align: center;
+    position: sticky;
+    top: 0;
+    z-index: 1;
+
+    .alert-banner {
+        margin: 0 auto;
+        display: flex;
+        justify-content: space-between;
+        border: 0;
+        border-radius: 0;
+    }
+
+    .bs2-deprecation-message-button {
+        font-size: 1.75rem;
+        right: 0;
+        top: 0;
+        transform: translateX(-100%);
+    }
+
+    #bs2-deprecation-message-content {
+        padding: 8px 0;
+        color: @red;
+        position: relative;
+        left: 50%;
+        transform: translateX(-50%);
+        margin-right: 65px;
+        font-size: 0.85rem;
+
+        a:not(.btn) {
+            text-shadow: none;
+            text-decoration: underline;
+        }
+    }
+}
```
Is all this style really needed using a Bootstrap 2 alert? Also, I think we use two spaces indentation in CSS.

anvit replied:
Some of this styling is just centering the content and aligning the close icon with the buttons from the privacy and update check notifications to be consistent, and some of it is to visually distinguish it from those two since there is a scenario where a user might see all three, and it made sense to make this appear more prominent than the other two. I could potentially combine some of the styles for all three of these under the same classes though and clean some of the duplicated code up (there's a little bit that's duplicated between the other two alerts already).

Good catch on the indentation though, it seems like my editor's indentation default was set to 4 instead of 2 spaces for CSS!

jraddaoui commented on apps/qubit/templates/_header.php:
```diff
@@ -1,3 +1,5 @@
+<?php echo get_component('default', 'bs2DeprecationMessage'); ?>
+
```
We should add this in arArchivesCanadaPlugin too: https://github.com/artefactual/atom/blob/qa/2.x/plugins/arArchivesCanadaPlugin/templates/_header.php#L2.

jraddaoui reviewed (changes requested):
Looking good @anvit!

I added some comments below for the current implementation but, as you noticed, what happens with users using a custom theme/header? My first thought is that we should inform them differently, maybe in the AtoM downloads page and of course in the release notes/announcement.

The reasoning behind that thought is that we could try to target some of them using a flash alert instead of an extra component/template, but we changed how those alerts are rendered in 2.7 and, if they are facing this issue, they are probably missing those flash alerts already.

So this will only affect users with a pretty basic theme or using the default B2 themes and it may be redundant if they already read the release notes/announcement, but it'll be a good reminder over time for them.

anvit commented:
Thanks so so much for the great feedback @jraddaoui (and for catching a bunch of issues!). Will add a commit fixing these. 

I just had another thought. Would it make things easier if I added this to the menu instead of the header (while having it still visually display where it is right now, as a sticky notification at the top of the header)? We're still going to do release announcements, so most users will hear about it that way, but I was wondering if that makes it automatically visible for folks using BS2 themes with custom headers?

jraddaoui commented:
Hard to tell @anvit, not sure what menu are you referring to, but I guess it could have been customized too. Another option following your idea would be to modify the response directly from a config/filter class adding the HTML code. Something similar to what we do [here](https://github.com/artefactual/atom/blob/qa/2.x/plugins/arDominionPlugin/config/arDominionPluginConfiguration.class.php#L32) but modifying the content and doing it from `apps/qubit/config/qubitConfiguration.class.php` or from a filter.

I think we are worrying too much, I don't like that idea a lot and a static alert can get messy to style properly. Give it a try if you want and think it's good solution, otherwise I'd leave it as it is. Nice work!

anvit commented:
Addressed feedback, and updated the URL to the page where the details about Bootstrap 2 deprecation and instructions for changing to the Bootstrap 5 Dominion theme are going to be. Also decided to keep the notice in the header since that'll be the easiest to manage and maintain in the future.

jraddaoui commented on apps/qubit/modules/default/actions/bs2DeprecationMessageComponent.class.php:
```diff
@@ -26,13 +26,14 @@ class DefaultBS2DeprecationMessageComponent extends sfComponent
-        if (!$this->context->user->isAuthenticated() || sfConfig::get('app_b5_theme', false)) {
-            return sfView::NONE;
-        }
+        $hasTranslateOrEditAccess = QubitAcl::check($this->resource, 'update')
```
I'm not sure if we should check against `$this->resource`. This component may be called from all pages, so I'm not sure if a resource will exist in all cases.

anvit replied:
Ah, you're right! I'm updating it to simply test against `QubitInformationObject` instead, which is something I'd seen in a couple other places [such as in the sfIsdiahPlugin](https://github.com/artefactual/atom/blob/c088704e63506d6bca5d9ea526e04fe59f12c5f9/plugins/arDominionB5Plugin/modules/sfIsdiahPlugin/templates/indexSuccess.php#L280). It seems to work, but let me know if you have a better suggestion!

jraddaoui replied:
Not sure how that would work, it kind of makes sense for creation. but not for update. I'd just remove it and trust on the groups (not sure if admin check will be needed to).

anvit replied:
Removing this breaks the check for admin accounts, so I'm adding the admin check in.

jraddaoui commented on js/bs2DeprecationMessage.js:
```diff
@@ -4,25 +4,24 @@
   $(function ()
   {
-    var $node = $('body');
+    let $node = $('#bs2-deprecation-message');
```
Nice.

jraddaoui commented on plugins/arArchivesCanadaPlugin/css/less/main.less:
```diff
@@ -319,3 +319,48 @@ body.login #content {
+      text-decoration: underline;
+    }
+  }
+}
```
arArchivesCanadaPlugin is extending arDominionPlugin, if the style is the same this shouldn't be needed.

anvit replied:
Fixed!

jraddaoui reviewed (changes requested):
Nicely done @anvit, thanks!

jraddaoui reviewed (approved):
Thanks @anvit!!
